File: j2kt/jre/java/native/java/lang/String.java
/*
 *  Licensed to the Apache Software Foundation (ASF) under one or more
 *  contributor license agreements.  See the NOTICE file distributed with
 *  this work for additional information regarding copyright ownership.
 *  The ASF licenses this file to You under the Apache License, Version 2.0
 *  (the "License"); you may not use this file except in compliance with
 *  the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package java.lang;

import static javaemul.internal.KtNativeUtils.ktNative;

import java.io.UnsupportedEncodingException;
import java.nio.charset.Charset;
import java.util.Comparator;
import java.util.Locale;
import javaemul.internal.annotations.KtName;
import javaemul.internal.annotations.KtNative;
import javaemul.internal.annotations.KtProperty;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

// TODO(b/223774683): Java String should implement Serializable. Kotlin String doesn't.
@KtNative(name = "kotlin.String", companionName = "java.lang.String")
@NullMarked
public final class String implements Comparable<String>, CharSequence {

  public static final Comparator<String> CASE_INSENSITIVE_ORDER = ktNative();

  public String() {}

  public String(byte[] data) {}

  @Deprecated
  public String(byte[] data, int high) {}

  public String(byte[] data, int offset, int byteCount) {}

  @Deprecated
  public String(byte[] data, int high, int offset, int byteCount) {}

  public String(byte[] data, int offset, int byteCount, String charsetName)
      throws UnsupportedEncodingException {}

  public String(byte[] data, String charsetName) throws UnsupportedEncodingException {}

  public String(byte[] data, int offset, int byteCount, Charset charset) {}

  public String(byte[] data, Charset charset) {}

  public String(char[] data) {}

  public String(char[] data, int offset, int charCount) {}

  public String(String toCopy) {}

  // TODO(b/224969395): Add once StringBuffer is available
  // public String(StringBuffer stringBuffer) {}

  public String(int[] codePoints, int offset, int count) {}

  public String(StringBuilder stringBuilder) {}

  @Override
  public native char charAt(int index);

  @Override
  public native int compareTo(String string);

  public native int compareToIgnoreCase(String string);

  @KtName("plus")
  public native String concat(String string);

  public static native String copyValueOf(char[] data);

  public static native String copyValueOf(char[] data, int start, int length);

  public native boolean endsWith(String suffix);

  @Override
  public native boolean equals(@Nullable Object object);

  public native boolean equalsIgnoreCase(@Nullable String string);

  @Deprecated
  public native void getBytes(int start, int end, byte[] data, int index);

  public native byte[] getBytes();

  public native byte[] getBytes(String charsetName) throws UnsupportedEncodingException;

  public native byte[] getBytes(Charset charset);

  public native void getChars(int start, int end, char[] buffer, int index);

  @Override
  public native int hashCode();

  public native int indexOf(int c);

  public native int indexOf(int c, int start);

  public native int indexOf(String string);

  public native int indexOf(String subString, int start);

  // TODO(b/222269323): Revisit.
  public native String intern();

  public native boolean isEmpty();

  public native int lastIndexOf(int c);

  public native int lastIndexOf(int c, int start);

  public native int lastIndexOf(String string);

  public native int lastIndexOf(String subString, int start);

  @Override
  @KtProperty
  public native int length();

  public native boolean regionMatches(int thisOffset, String other, int otherOffset, int length);

  public native boolean regionMatches(
      boolean ignoreCase, int thisOffset, String other, int otherOffset, int length);

  public native String repeat(int count);

  public native String replace(char oldChar, char newChar);

  @KtName("java_replace")
  public native String replace(CharSequence target, CharSequence replacement);

  public native boolean startsWith(String prefix);

  public native boolean startsWith(String prefix, int start);

  public native String substring(int start);

  public native String substring(int start, int end);

  public native char[] toCharArray();

  @KtName("lowercase")
  public native String toLowerCase();

  public native String toLowerCase(Locale locale);

  @Override
  public native String toString();

  @KtName("uppercase")
  public native String toUpperCase();

  public native String toUpperCase(Locale locale);

  @KtName("java_trim")
  public native String trim();

  public static native String valueOf(char[] data);

  public static native String valueOf(char[] data, int start, int length);

  public static native String valueOf(char value);

  public static native String valueOf(double value);

  public static native String valueOf(float value);

  public static native String valueOf(int value);

  public static native String valueOf(long value);

  public static native String valueOf(@Nullable Object value);

  public static native String valueOf(boolean value);

  // TODO(b/224969395): Add once StringBuffer is available
  // public native boolean contentEquals(StringBuffer strbuf);

  public native boolean contentEquals(CharSequence cs);

  @KtName("java_matches")
  public native boolean matches(String regularExpression);

  public native String replaceAll(String regularExpression, String replacement);

  public native String replaceFirst(String regularExpression, String replacement);

  public native String[] split(String regularExpression);

  public native String[] split(String regularExpression, int limit);

  @Override
  public native CharSequence subSequence(int start, int end);

  public native int codePointAt(int index);

  public native int codePointBefore(int index);

  public native int codePointCount(int beginIndex, int endIndex);

  public native boolean contains(CharSequence cs);

  public native int offsetByCodePoints(int index, int codePointOffset);

  public static native String format(String format, @Nullable Object... args);

  public static native String format(
      @Nullable Locale locale, String format, @Nullable Object... args);

  public static native String join(
      CharSequence delimiter, Iterable<? extends @Nullable CharSequence> elements);

  public static native String join(CharSequence delimiter, @Nullable CharSequence... elements);
}

File: j2kt/jre/java/native/java/lang/String.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package java.lang

import java.io.UnsupportedEncodingException
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.nio.charset.UnsupportedCharsetException
import java.util.Formatter
import java.util.Locale
import java.util.StringJoiner
import javaemul.lang.appendCodePoint
import kotlin.collections.Iterable as KotlinIterable

/**
 * Pseudo-constructor for emulated java.lang.String.
 *
 * We assume that the platform encoding is UTF-8 here (like J2Objc and Android do).
 *
 * See regular JRE API documentation for other methods in this file.
 */
object String {
  val CASE_INSENSITIVE_ORDER: Comparator<kotlin.String> = kotlin.String.CASE_INSENSITIVE_ORDER

  operator fun invoke(a: CharArray): kotlin.String = valueOf(a)

  operator fun invoke(sb: StringBuilder): kotlin.String = sb.toString()

  operator fun invoke(a: CharArray, offset: Int, len: Int): kotlin.String = valueOf(a, offset, len)

  operator fun invoke(a: ByteArray) = a.decodeToString(throwOnInvalidSequence = false)

  operator fun invoke(a: ByteArray, offset: Int, len: Int) =
    a.decodeToString(offset, offset + len, throwOnInvalidSequence = false)

  operator fun invoke(a: IntArray, offset: Int, len: Int): kotlin.String {
    val sb = StringBuilder(len)
    for (i in offset until offset + len) {
      sb.appendCodePoint(a[i])
    }
    return sb.toString()
  }

  operator fun invoke(a: ByteArray, offset: Int, len: Int, charSet: Charset): kotlin.String =
    when (charSet) {
      StandardCharsets.UTF_8 ->
        a.decodeToString(offset, offset + len, throwOnInvalidSequence = false)
      StandardCharsets.US_ASCII -> a.decodeToStringUnmapped(offset, offset + len, ascii = true)
      StandardCharsets.ISO_8859_1 -> a.decodeToStringUnmapped(offset, offset + len, ascii = false)
      else -> throw UnsupportedCharsetException(charSet.name())
    }

  operator fun invoke(a: ByteArray, charSet: Charset) = String(a, 0, a.size, charSet)

  operator fun invoke(a: ByteArray, charSetName: kotlin.String) = String(a, 0, a.size, charSetName)

  operator fun invoke(
    a: ByteArray,
    offset: Int,
    len: Int,
    charsetName: kotlin.String,
  ): kotlin.String {
    // In Java, UnsupportedCharsetException is an unchecked exception thrown by Charset.forName;
    // Methods taking a character set name here are expected to throw the checked
    // UnsupportedEncodingException instead.
    try {
      return String(a, offset, len, Charset.forName(charsetName))
    } catch (e: UnsupportedCharsetException) {
      throw UnsupportedEncodingException(charsetName)
    }
  }

  operator fun invoke(s: kotlin.String) = StringBuilder(s).toString()

  fun format(format: kotlin.String, vararg values: Any?): kotlin.String =
    Formatter().format(format, *values).toString()

  fun format(locale: Locale?, format: kotlin.String, vararg values: Any?): kotlin.String =
    Formatter().format(locale, format, *values).toString()

  fun join(delimiter: CharSequence, elements: KotlinIterable<CharSequence>): kotlin.String {
    val joiner = StringJoiner(delimiter)
    for (element in elements) {
      joiner.add(element)
    }
    return joiner.toString()
  }

  fun join(delimiter: CharSequence, vararg elements: CharSequence): kotlin.String =
    join(delimiter, elements.asList())

  fun valueOf(c: Char): kotlin.String = c.toString()

  fun valueOf(a: Any?) = if (a is Int) Integer.toString(a) else a.toString()

  fun valueOf(i: Int): kotlin.String = Integer.toString(i)

  fun valueOf(data: CharArray): kotlin.String = data.concatToString()

  fun valueOf(data: CharArray, offset: Int, count: Int) =
    data.concatToString(offset, offset + count)

  fun copyValueOf(data: CharArray) = valueOf(data)

  fun copyValueOf(data: CharArray, offset: Int, count: Int): kotlin.String =
    valueOf(data, offset, count)

  private fun ByteArray.decodeToStringUnmapped(
    offset: Int,
    end: Int,
    ascii: kotlin.Boolean,
  ): kotlin.String {
    val sb: StringBuilder = StringBuilder(end - offset)
    for (i in offset until end) {
      val c = this[i].toInt() and 255
      if (ascii && c > 127) {
        throw CharacterCodingException()
      }
      sb.append(c.toChar())
    }
    return sb.toString()
  }
}
